internal/propagation.go

```diff
@@ -0,0 +1,115 @@
+package internal
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// assumes a header of the form:
+
+// VERSION;PAYLOAD
+
+// VERSION=1
+// =========
+// PAYLOAD is a list of comma-separated params (k=v pairs), with no spaces.  recognized
+// keys + value types:
+//
+//  trace_id=${traceId}    - traceId is an opaque ascii string which shall not include ','
+//  parent_id=${spanId}    - spanId is an opaque ascii string which shall not include ','
+//  context=${contextBlob} - contextBlob is a base64 encoded json object.
+//
+// ex: X-Honeycomb-Trace: 1;trace_id=weofijwoeifj,parent_id=owefjoweifj,context=SGVsbG8gV29ybGQ=
+
+const (
+	TracePropagationHTTPHeader = "X-Honeycomb-Trace"
+	TracePropagationVersion    = 1
+)
+
+type Propagation struct {
+	TraceHeader
+	TraceContext       map[string]interface{}
+	TraceContextBase64 string
+}
+
+type PropagationError struct {
+	message      string
+	wrappedError error
+}
+
+func (p *PropagationError) Error() string {
+	if p.wrappedError == nil {
+		return p.message
+	}
+	return fmt.Sprintf(p.message, p.wrappedError)
+}
+
+func MarshalTraceContext(ctx context.Context) string {
```

Comment:
Might be nice to have `MarshalTraceContext`/`UnmarshalTraceContext` available to consumers of the public API. So that if someone needs to they can take the trace context and stuff it in a protobuf, or otherwise manage the propagation themselves.